### PR TITLE
documentSymbol / outline support

### PR DESCRIPTION
**Pull request by rlivings39**
Here's an initial attempt at outline support for https://github.com/mathworks/MATLAB-extension-for-vscode/issues/1

This has a few responsiveness bugs I need to investigate, so please don't actually merge yet. Still I wanted to get some initial feedback.

Some bits of feedback I'd like to hear about:

* Is this the right location? It feels navigation-related so I put it there. I could also imagine making a separate provider folder like `symbols` or `structure`.
* I've tried to integrate well with the MATLAB startup option. Editors make requests for document symbols very frequently so I chose to make the document symbol request not trigger a MATLAB start with `onDemand` as mentioned in my comment.
* One UX issue I didn't immediately see how to solve is: suppose you have A.m and foo.m open at VSCode launch. The documentSymbol request fires and since MATLAB hasn't started, we don't have any code index info yet. Right now I just return empty. It would be nice if I could wait on a promise or something for files to be indexed and then immediately return a response once the system is up and running. Any ideas on whether or not there's anything in the existing server that I could leverage for this?

Here's a screenshot of this in action:

image

Things left to do

- [x] Fix handling of classdef folders
- [x] Look into apparent stale data showing up with classdefs
- [x] Look at responsiveness for already opened files. Wait on MATLAB connection. Ensure promises don't stack up.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
         "target": "ES6",
         "rootDir": "./src",
         "outDir": "./out",
+		"sourceMap": true,
```
Comment (dklilley) [inline tsconfig.json:7]: Use spaces, not tabs

**Comment by dklilley** [comment]
Can you also add an entry to the list in the README for document symbol support?

**Comment by zm-cttae** [comment]
Does this work with mpm.m (multiple functions) and also a datastorage.m hugefile? Is there:

- variable assignment symbols
- anonymous function support etcetc

File: src/providers/navigation/NavigationSupportProvider.ts [hunk 2]
```diff
@@ -113,6 +114,62 @@ class NavigationSupportProvider {
...
+         */
+        function pushSymbol(name: string, kind: SymbolKind, symbolRange: Range) {
+            if (!visitedRanges.has(symbolRange)) {
+                result.push(SymbolInformation.create(name, kind, symbolRange, uri));
+                visitedRanges.add(symbolRange);
+            }
+        }
+        if (codeData.isClassDef && codeData.classInfo != null) {
+            let classInfo = codeData.classInfo;
+            // TODO: When can this be null if we're in a classInfo case?
```
[the remark this answers is not in the document]
Comment (rlivings39) [reply src/providers/navigation/NavigationSupportProvider.ts:173]: Got it. I need to handle this case and class folders in general better. Will fix that up.

File: tsconfig.json (hunk 1 above)
Comment (rlivings39) [reply tsconfig.json]: Fixed

File: src/providers/navigation/NavigationSupportProvider.ts [hunk 3]
```diff
@@ -59,7 +59,8 @@ class Expression {
 
 export enum RequestType {
     Definition,
-    References
+    References,
+    DocumentSymbol
 }
 
 function reportTelemetry (type: RequestType, errorCondition = '') {
```
[the remark this answers is not in the document]
Comment (rlivings39) [reply src/providers/navigation/NavigationSupportProvider.ts]: Fixed. Let me know what you think.

**Comment by rlivings39** [comment]
> Can you also add an entry to the list in the README for document symbol support?

Done

**Comment by zm-cttae** [comment]
Links to nice test cases:

- https://github.com/mobeets/mpm/blob/90a42c39365a1b508e9f143f1f562c89e14edeb1/mpm.m
- https://github.com/jgte/orb/blob/df184892563f22e449db85c8ebb09a9315c688f2/datastorage.m

To start with, the loops listed here: [*Loops and Conditional Statements - MATLAB & Simulink - MathWorks United Kingdom*](https://uk.mathworks.com/help/matlab/control-flow.html)

**Review by dklilley — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -19,4 +20,4 @@ MATLAB language server supports these editors by installing the corresponding ex
 ### 1.0.0
```
[the remark this answers is not in the document]
Comment (rlivings39) [reply README.md:25]: Done

File: src/providers/navigation/NavigationSupportProvider.ts [hunk 5]
```diff
@@ -113,6 +126,67 @@ class NavigationSupportProvider {
...
+
+        if (textDocument == null) {
+            reportTelemetry(requestType, 'No document')
+            return []
+        }
+        await Indexer.indexDocument(textDocument)
+        let codeData = FileInfoIndex.codeDataCache.get(uri)
+        if (codeData === null) {
+            // Ask to index file
+            await Indexer.indexDocument(textDocument)
```
[the remark this answers is not in the document]
Comment (rlivings39) [reply src/providers/navigation/NavigationSupportProvider.ts]: Nope. Thanks for catching this.

**Review by dklilley — APPROVED** [review]

**Pull request merged**